### ISSUE TITLE
DNS requests might bypass Adguard's tunnel in some circumstances

### ISSUE THREAD

**Issue by jo2dan23** [AdguardTeam/AdguardForiOS#404]
I know there is another issue on this but I think this one is different because it completely stops logging domains in my dns log also. I have to turn vpn on and off then it works again. What file would show whats going on? Im suprised that im like the only one who's experiencing this.

**Comment by jo2dan23** [comment]
Odd thing i noticed. I think this only happens on mobile network

**Comment by ameshkov** [comment]
@jo2dan23 what's your cellular provider? I wonder, could it be the same issue as we have with Android version: https://github.com/AdguardTeam/AdguardForAndroid/issues/1026

**Comment by jo2dan23** [comment]
That could be the issue. Alot of unblocked ads i see are coming only when I use my data. I have cellcom in Wisconsin

**Comment by ameshkov** [comment]
And this all started with the new update where we added IPv6 routes. I wonder if it's the case.

@IvanIin @Stillness-2 guys, please take a look

**Comment by ameshkov** [comment]
My idea is the following:

If we add an IPv6 interface, we should add routes for IPv4-mapped addresses of the DNS servers.

For instance, if DNS server address is `8.8.8.8`, **and** we added an IPv6 interface, we should add two routes:

`8.8.8.8/32` -- for IPv4 interface
`::FFFF:8.8.8.8/32` -- for IPv6 interface (address syntax is wrong, just for better understanding)

**Comment by IvanIin** [comment]
@jo2dan23  can you please sent us logs from the latest beta version of the adGuard? With system-wide ON and System Default DNS enabled. And tell the time in which the issue was observed.

**Comment by ameshkov** [comment]
Let me please explain, we've improved logging in the latest beta, added network interfaces logging, it might help to understand what's wrong.

**Comment by jo2dan23** [comment]
Which logs do you need? Oddly i cannot make it show me ads on data at the moment.

**Comment by ameshkov** [comment]
We need tunnel log

**Comment by jo2dan23** [comment]
I grabbed these the days the ads were shown on mobile network hope they help.

https://1drv.ms/u/s!AvccVnEjavPPhxJd95Xovra_We8k

https://1drv.ms/u/s!AvccVnEjavPPhxP1hmTxLajIccwU

**Comment by zebrum** [comment]
ticket id 1570161

Ads were showing in apps even though System Wide protection was on.

Noticed the DNS requests log wasn't logging entries even though it was enabled.

To resolve it, I need to manually toggle System Wide protection off and on for ads to stop showing, and for the requests log to populate.

Ads were showing at 2:50pm, then I manually turned protection off and on and successfully blocked ads at 2:51pm onwards.



[Archive.zip](https://github.com/AdguardTeam/AdguardForiOS/files/1054439/Archive.zip)

**Comment by ameshkov** [comment]
So, after all, it appears that in some circumstances DNS requests bypass Adguard's tunnel. We need to extend AG logging in order to find out what's going on. New beta will be pushed today.

**Comment by jo2dan23** [comment]
@ameshkov it is weird this only occurs on mobile networks though? And i think it did start on the ipv6 version cant remember though.

**Comment by planet0** [comment]
I was the one who also reported this issue and submitted ticket id 1570161 as @zebrum posted.
@ameshkov, happy to send this when the issue reoccurs after recieving the new beta.
@jo2dan23, this was occuring on WiFi networks for me as well as mobile networks.

**Comment by jo2dan23** [comment]
@planet0 glad im not the only one here. Its odd this doesnt happen to me on wifi. Could it be that my isp supports ipv6 or does that not have anything to do with this? @ameshkov

**Comment by jo2dan23** [comment]
@planet0 what apps does it do this in? Which app is the most easily reproducable?

**Comment by planet0** [comment]
@jo2dan23 Hmm, both my mobile network and Wi-Fi use IPV6, does your Wi-Fi use IPV6?

I found it's easiest to notice with apps that show an obvious ad banner when launched (then use the app drawer to close completely and start the process again). The two I use are Speedtest by Ookla (banner shown after tapping 'Start Test' and waiting until it tests download speed) and Splasher by Adam Tootle (after scrolling down several images in the grid).

**Comment by jo2dan23** [comment]
@planet0 well i get an ipv6 address shown on my router and in wifi settings on my iphone. Whisper is one app I use and meetme is the other. Only show ads on mobile though so i don't understand

**Comment by jo2dan23** [comment]
I just used a website to test ipv6. Looks like both cellular and wifi have it. So i really don't understand. Lol

**Comment by planet0** [comment]
@jo2dan23, might not be related to IPV6 then.

I'll also mention that this issue almost always occurs when I take my iPhone off Aeroplane Mode first thing in the morning (as I turn Aeroplane Mode on when I go to sleep the night before). The device will reconnect to my Wi-Fi and then display the VPN icon, and I then wait at least 60 seconds before opening apps. This is when I notice the issue. Once I manually turn System Protection off and back on, the issue is temporarily resolved.

I've also experienced this issue if I haven't used my iPhone for long periods of time.

**Comment by jo2dan23** [comment]
ive had it happen multiple times a couple days ago on data. It was a stable connection to towers so there was no disconnect and reconnect  of the vpn at all. So idk i hope this can be resolved soon.

**Comment by jo2dan23** [comment]
@planet0 were you able to reproduce today on latest beta?

**Comment by planet0** [comment]
@jo2dan23, not exactly (see below).
@ameshkov, this might be another circumstance for bypassing.

Something different I had enabled this time (on by default on iOS devices) was Wi-Fi assist which would automatically switch to mobile data if the Wi-Fi wasn’t strong enough.

In this instance, once I took Aeroplane Mode off this morning, my iPhone was switching between Wi-Fi and Mobile Data every 5 seconds of so, which was not good for the VPN either. Ads were bypassing freely whilst this was happening. This may be expected behaviour, but this isn’t good if the vast majority of iOS users will be in this scenario due to Wi-Fi assist, and this causes the VPN not to work. Was really strange, and I’ve turned this Wi-Fi assist feature off from now on.

I’ve sent a bug report via the app and noted the time of bypassing (4:41am), once again my user ID is 1570161.

I’ll test again with Wi-Fi assist turned off and see what happens.

**Comment by jo2dan23** [comment]
@planet0 that's probably why i dont see ads on wifi i turned that option off a long time ago. Makes sense now for that atleast.

**Comment by planet0** [comment]
@jo2dan23 I only had the option on for less than 12 hours and noted the results above, ads were bypassing for me when it was off (like my first report).

**Comment by jo2dan23** [comment]
@planet0 i think the reason why i dont see ads on wifi might be cuz i have mostly all the domains in my blacklist that are in the exceptions list. So one of those could be causing ads.

**Comment by jo2dan23** [comment]
@ameshkov there is a bug in latest beta my wifi and vpn keep disconnecting. And i dont have wifi assist on.

**Comment by X8716e** [comment]
@ameshkov --- You likely already tried doing this, though this is what I did to get it to connect quickly & stay connected: Delete Adguard Pro app from your home screen. Reinstall via TestFlight. Setup per standard instructions. I hope this helps you.

**Comment by jo2dan23** [comment]
@X8716e i think that worked thanks. @ameshkov i just got the ads. Finally. Here is my log. I missed the time it occurred so just look between 6pm and 615pm https://1drv.ms/u/s!AvccVnEjavPPhxljifrCb1Dk36y5

**Comment by jo2dan23** [comment]
Just captured another log. 650pm - 654pm ads appear 

https://1drv.ms/u/s!AvccVnEjavPPhxpw_5Lcfpjp2fAX

**Comment by jo2dan23** [comment]
@ameshkov ok so it appears blocking stops completely thats why ads appear. All the domains in my blocklist arent being blocked.

**Comment by jo2dan23** [comment]
@ameshkov looking at my tunnel logs now... there is nothing but error codes?  no domains being logged could it be problems with using the opendns server? i dont think enabling the logging in the latest beta even did anything

**Comment by jo2dan23** [comment]
Well now its doing in on wifi :( switched back to latest stable and doing it also in there. This is frustrating

**Comment by jo2dan23** [comment]
Here is the log I captured on latest stable version. Wonder if it helps. Ads appear between 2pm and 205pm

https://1drv.ms/u/s!AvccVnEjavPPhxvirVXClLp2tcoK

**Comment by ameshkov** [comment]
Sorry for the late reply, guys.

@jo2dan23 
> @planet0 glad im not the only one here. Its odd this doesnt happen to me on wifi. Could it be that my isp supports ipv6 or does that not have anything to do with this? @ameshkov

Frankly, I now don't think IPv6 is related directly. It is more like that iOS is trying to be too smart and starts bypassing tunnel if some of the domain names were not resolved in time.

**Comment by jo2dan23** [comment]
@ameshkov my logs only show errors sorry i couldnt get a log that did anything to help. So basically the phone probably needs to be plugged into computer and be debugged? Or whats that program where it logs everything the phone is doing. Hope this is fixable.

**Comment by ameshkov** [comment]
Yeah, iOS logs can be helpful indeed:
https://github.com/AdguardTeam/AdguardForiOS/issues/384#issuecomment-297385621

**Comment by jo2dan23** [comment]
@ameshkov heres the weird thing tho. I literally cannot get ads to appear on mobile network while at home. Which I'm guessing like you said is because if the domains arent resolved fast enough they bypass the vpn. Which my tower by home is fast. It only happens when im traveling and the cell towers arent good and data is slow.

**Comment by planet0** [comment]
Just an update — I have not been able to reproduce the issue since disabling IPV6 on my Wi-Fi network (Needed to do this in order to use Adguard DNS as there is no IPV6 support yet and I cannot bypass ISP assigned DNS with IPV6 enabled), and this issue has not reoccurred on mobile data either (I was previously incorrect about IPV6, mobile data only uses IPV4).

**Comment by jo2dan23** [comment]
I am on ios 11. And havent seen ads appear yet on mobile network. Though it might be because some frameworks are broken in the apps since they aren’t updated for ios 11 yet

**Comment by jo2dan23** [comment]
But i am on the stable version of adguard pro havent tried beta because I think the logging that was enabled overloads ios memory and makes vpn disconnect

**Comment by planet0** [comment]
So far so good with build 1.2.2 (88)!

**Comment by ameshkov** [comment]
> So far so good with build 1.2.2 (88)!

Fingers crossed!

**Comment by jo2dan23** [comment]
@ameshkov. I think it might be fixed but ads are showing again in certain apps because googleadservices was added back to the exceptions list and i still cant override it using blacklist. So i cant be 100 percent sure if its fixed

**Comment by jo2dan23** [comment]
Good news. Im pretty sure this is resolved now. Been using my data for about 20 mins and no ads yet

**Comment by ameshkov** [comment]
@IvanIin @Stillness-2

Let's have a recap.

The latest beta had two changes:
1. System DNS was added to routes just in case if iOS decides to ignore AG DNS
2. We've extended logging (which might've caused the slow down).

We should resolve the slowdown issue @jo2dan23 is talking about and we're good to go.

**Comment by jo2dan23** [comment]
Should we close this now?

**Comment by Stillness-2** [comment]
no

**Comment by ameshkov** [comment]
> Should we close this now?

We'd better wait for the next beta and make sure that the new approach resolves it.

**Comment by IvanIin** [comment]
https://developer.apple.com/documentation/networkextension/nevpnmanager

`If both a Personal VPN configuration and a Non-Personal VPN configuration are connected simultaneously, and both VPN tunnels are configured to act as the default route that network traffic takes to reach the Internet, then the Non-Personal VPN tunnel will actually be the default route to the Internet for as long as it is connected.`

We use Non-Personal VPN configuration. If we add default routes to included routes, then other Personal VPN tunnels will stop working (Non-Personal tunnels never work simultaneously). 

We can add "full-tunnel(no other VPN)/split-tunnel/automatic" switch to the advanced settings.

**Comment by ameshkov** [comment]
My suggestion:

1. Introduce a new advanced setting called "Tunnel mode"
2. There will be three possible values: `Auto`, `Split-Tunnel`, `Full-Tunnel`

* `Split-Tunnel` -- if AdGuard works in this mode, it is compatible with "Personal VPN" apps. List of VPN apps compatibility: #162. The downside is that in the case of bad connectivity, iOS may bypass AdGuard.
* `Full-Tunnel` -- if AdGuard works in this mode, it is incompatible with other VPN apps. On the other hand, it will be used regardless of the connection quality.
* `Auto` - AdGuard will use `Split-Tunnel` mode if an active VPN is detected. Otherwise, it will operate in the `Full-Tunnel` mode.

**Comment by jo2dan23** [comment]
Bad news. Ads are back on mobile network

**Comment by jo2dan23** [comment]
https://1drv.ms/u/s!AvccVnEjavPPh08alBsrFRZu_WdB

**Comment by jo2dan23** [comment]
Does my log help? Hopefully:/ maybe its my carrier bypassing the vpn somehow?

**Comment by ameshkov** [comment]
@jo2dan23 the new `Full-Tunnel` mode should resolve: https://github.com/AdguardTeam/AdguardForiOS/issues/404#issuecomment-309760194

**Comment by jo2dan23** [comment]
Ohh. So the reason its doing that is because its not fully staying enabled? I dont use another vpn alongside it tho.

**Comment by ameshkov** [comment]
> Ohh. So the reason its doing that is because its not fully staying enabled? I dont use another vpn alongside it tho.

Yeah, the current implementation operates in the `Split-Tunnel` mode so it has all the drawbacks described.

**Comment by jo2dan23** [comment]
Awesome! What settings do i enter to make opendns work on mobile?

**Comment by jo2dan23** [comment]
Eta on next beta? Sorry to bother

**Comment by ameshkov** [comment]
@jo2dan23 

> Eta on next beta? Sorry to bother

Friday

> What settings do i enter to make opendns work on mobile?

It should work with default settings.

**Comment by jo2dan23** [comment]
I know I asked this already but will i be able to block googleadservices in next release? I cant test the ad problem on mobile network without blocking that domain the app i use to test appears ads when google ads is unblocked.

**Comment by jo2dan23** [comment]
This one: https://github.com/AdguardTeam/AdguardForiOS/issues/354

**Comment by ameshkov** [comment]
Here is an update on this issue.

It appears, that implementing an `Auto` mode as I suggested [here](https://github.com/AdguardTeam/AdguardForiOS/issues/404#issuecomment-309760194) is almost impossible due to how VPN tunnels work.

So, we will provide two options:

1. `Split-Tunnel` - default mode, compatible with third-party VPN services. A bit improved so that there was a lower probability of iOS bypassing AG.
2. `Full-Tunnel` - in this mode AdGuard works as a primary network interface, hence incompatible with other VPN apps. However, in this mode, nothing can make iOS bypass AG.

**Comment by planet0** [comment]
Sounds like a great idea to resolve this issue, I'd set it to Full-Tunnel even though I don't use any other VPN.

**Comment by jtb1234** [comment]
I was directed here from the Adguard for IOS technical help forum.

My issue is the opposite. A couple days ago I started getting ads in apps while on wifi. Apps are slow loading content such as Letgo and offerup. Web pages in Safari are slow to load.

**Comment by ameshkov** [comment]
> Sounds like a great idea to resolve this issue, I'd set it to Full-Tunnel even though I don't use any other VPN.

Finally available in the beta build 92.

> My issue is the opposite. A couple days ago I started getting ads in apps while on wifi. Apps are slow loading content such as Letgo and offerup. Web pages in Safari are slow to load.

As I understand, you may be suffering from the very same issue. Want to try the latest beta version and check if it's resolved there?

**Comment by planet0** [comment]
> Finally available in the beta build 92.

Really happy with this, and not seeing the VPN icon is a big bonus for me.
Not having any issues here @ameshkov 👍🏼

**Comment by jtb1234** [comment]
Yes I would like to give the latest beta a try.

**Comment by ameshkov** [comment]
@planet0 
> Really happy with this, and not seeing the VPN icon is a big bonus for me.

Btw, it seems that this is an iOS bug. Not sure should we report it or consider it an undocumented feature and live with it:)

**Comment by jo2dan23** [comment]
Unfortunately it seems this bug is still not fixed. I think it may have got worse. Now ads appear on wifi :(

**Comment by jo2dan23** [comment]
I couldnt capture a log because it happened when logging was off..

**Comment by jo2dan23** [comment]
Well i havent see anymore ads. So idk. Should i close?

**Comment by ameshkov** [comment]
> Well i havent see anymore ads. So idk. Should i close?

Let's keep it open for some time, unless we're 100% sure.

**Comment by jtb1234** [comment]
> As I understand, you may be suffering from the very same issue. Want to try the latest beta version and check if it's resolved there?

Yes I would like to try out the beta.

**Comment by vozersky** [comment]
@jtb1234  pls send me your apple id through pm on forum and I'll send you the testflight key https://forum.adguard.com/index.php?conversations/add&to=aegis

**Comment by jtb1234** [comment]
@vozersky 

> pls send me your apple id through pm on forum and I'll send you the testflight key

Info sent

**Comment by vozersky** [comment]
@jtb1234 check your pm :)

**Comment by jtb1234** [comment]
Installed the beta and everything seeems fine so far. Ads blocked system wide.

**Comment by jtb1234** [comment]
I sent a bug report thru the app and posting it here also.

After I upgraded the app, apps like wunderground, weatherbug will not load data in Full Tunnel mode.

If I switch to Split Tunnel, the apps load data but there are ads.

**Comment by zebrum** [comment]
@jtb1234 can't reproduce with 1.2.2(95)

**Comment by jtb1234** [comment]
I just rebooted and tried again and get the same results. No app content or no web pages loading with full tunnel on wifi. The beta before the update was working fine.

**Comment by jo2dan23** [comment]
On latest with full tunnel mode. Wifi works for me. Go to your network settings and try a reset

**Comment by jtb1234** [comment]
Tried resetting network settings and didn't help. It shuts down all data it seems cause nothing loads.

**Comment by jtb1234** [comment]
I sent the logs. Like I said, when in full tunnel mode, I can't connect to anything needing data.

**Comment by ameshkov** [comment]
> Like I said, when in full tunnel mode, I can't connect to anything needing data.

@jtb1234 could you please check how it works when DNS is set to "System Default"? According to the log file, it seems that there's no access to AG DNS.

**Comment by jtb1234** [comment]
I switched to system default and everything seems to work as it should.

**Comment by ameshkov** [comment]
@jtb1234 What's your ISP? it seems that for some reason it does not allow using third-party DNS servers.

**Comment by ameshkov** [comment]
The problem revealed itself in the Full-tunnel mode as in split-tunnel iOS automatically switches to the ISP server in that case.

**Comment by jtb1234** [comment]
Brandenburg Telephone Company. It's a small local company.

**Comment by ameshkov** [comment]
> would that be a recent change. AG had been working up to a few days ago. It happened out of nowhere.

It's just @IvanIin fixed some of the issues in Full-Tunnel mode due to which ISP's DNS server might've been used in the previous builds. Now it operates as it should.

**Comment by jtb1234** [comment]
So what do you recommend now?

**Comment by ameshkov** [comment]
@jtb1234 at the moment - don't use AG DNS. Regardless of what mode you use it cannot be accessed. Just stick to the system-wide ad blocking.

**Comment by ameshkov** [comment]
@jtb1234 you may also report this situation to Brandenburg Telephone Company, check if this behavior is intended.

**Comment by jtb1234** [comment]
Just got off the phone with BBTEL and they said they shouldn't be blocking anything on their end.

**Comment by ameshkov** [comment]
> Just got off the phone with BBTEL and they said they shouldn't be blocking anything on their end.

Just in case, is it your broadband ISP or cellular? The problem is when you're on wi-fi.

**Comment by jtb1234** [comment]
It's a DSL connection.

**Comment by ameshkov** [comment]
@jtb1234 let's move this discussion to a new issue:
https://github.com/AdguardTeam/AdguardForiOS/issues/439
